src/samVisualize/graphGeneration/cfnTemplateGraphGenerator.ts

```diff
@@ -0,0 +1,135 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import { Graph, GraphObject } from './graph'
+import { yamlParse } from 'yaml-cfn'
+import { LinkTypes } from '../samVisualizeTypes'
+import * as _ from 'lodash'
+import { getLogger } from '../../shared/logger/logger'
+
+/**
+ * @param inputYaml A string representing a YAML template
+ * @returns A JavaScript object corresponding to the given YAML string
+ */
+function yamlStringToObject(inputYaml: string): Record<string, unknown> {
+    let templateAsObject = {}
+    try {
+        templateAsObject = yamlParse(inputYaml)
+    } catch (err) {
+        getLogger().error(
+            `SAM Visualize: Failed to load template "${inputYaml.substr(
+                0,
+                25
+            )}"... Please ensure the template is valid YAML: 0%`,
+            err
+        )
+    }
+    return templateAsObject
+}
+
+/**
+ * @description Takes a string with embedded resource names within `${}`, and extracts each resource.
+ * See https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-sub.html#w2ab1c33c28c59c11b6
+ * @example 'xxx${resource1}yyyzzz${resource2}abc' => ['resource1', 'resource2']
+ * @param value A string containing embedded resources
+ * @returns A list of strings corresponding to the embedded resource names. If no resource names are found, returns an empty list.
+ */
+function extractSubstitution(value: string): Array<string> {
+    const names: Array<string> | null = value.match(/\${[^}]+}/g)
```

Comment:
Can you do this with a capture group so you don't have to do a second regex?

Reply:
Refactored this. It's clearer, and only using a single regex :)